### ISSUE TITLE
Attach email to meeting invite

### ISSUE THREAD

**Issue by rjwiii** [ecederstrand/exchangelib#239]
Ok, so here's what I'm trying to do: I check a certain mailbox for new messages. I use information in those e-mails to generate meeting inviteas. I would now like to attach the original email to the meeting invite I'm creating. Is this possible?

From what I was able to gather from Google searches, it looks like I might have to save the original email locally and then attach it to the meeting invite. Question is: how? How do I save the original email in .msg format? What type do I use to attach it?

Any ideas?

Thanks,

rjwiii

**Comment by ecederstrand** [comment]
Attaching existing items in Exchange to other items in Exchange (e.g. attaching a `Message` to a `CalendarItem`) is actually quite straight-forward. See https://github.com/ecederstrand/exchangelib#attachments

If you can create a meeting invitation via exchangelib, then just use the `attach()` method and call it with the original `Message` object. This will attach the message when you call `save()` on the invitation. Beware, though, that meeting invitations are untested and quite poorly supported, so there's a possibility that creating a meeting invitation does not work at all. See https://github.com/ecederstrand/exchangelib/issues/89

**Comment by rjwiii** [comment]
Creating meeting invites is going fine for me so far (knock on wood).

So, to attach a Message item to my Calendar Item, would I just do something like:

my_msg_attach = exchangelib.ItemAttachmemt(my_msg_item)
my_appt.attach(my_msg_attach)

??

Also, while I'm thinking about it ... do you know of a way to send meeting invites that do not require a response (ie: clicking on "Accept" doesn't produce a drop down menu and/or message says "The organizer has not requested a response" at the top)?

Thanks,

rjwiii

**Comment by ecederstrand** [comment]
Yes, that's how it works.

I have no idea how the information on whether a response is required is sent. You could try to send one of those to yourself, and inspect the item afterwards. Although it's possible that the information is hidden in an extended property somewhere.

**Comment by rjwiii** [comment]
Following your docs, I have the following in my code:

vcn_appt.attach(exchangelib.ItemAttachment(name=vcn_email.subject, item=vcn_email))
vcn_appt.save(send_meeting_invitations=exchangelib.items.SEND_TO_ALL_AND_SAVE_COPY)

vcn_email is an item I pulled out from inbox.all().

I'm getting the following traceback:

Traceback (most recent call last):
  File "C:/Users/M82129/Git/Vendor Circut Notification/vcnmain.py", line 298, in <module>
    main(vcn_acct)
  File "C:/Users/M82129/Git/Vendor Circut Notification/vcnmain.py", line 281, in main
    vcn_appt.save(send_meeting_invitations=exchangelib.items.SEND_TO_ALL_AND_SAVE_COPY)
  File "C:\Python27\lib\site-packages\exchangelib\items.py", line 175, in save
    item = self._create(message_disposition=SAVE_ONLY, send_meeting_invitations=send_meeting_invitations)
  File "C:\Python27\lib\site-packages\exchangelib\items.py", line 189, in _create
    send_meeting_invitations=send_meeting_invitations)
  File "C:\Python27\lib\site-packages\exchangelib\account.py", line 266, in bulk_create
    send_meeting_invitations=send_meeting_invitations,
  File "C:\Python27\lib\site-packages\exchangelib\account.py", line 260, in <genexpr>
    i if isinstance(i, Exception)
  File "C:\Python27\lib\site-packages\exchangelib\services.py", line 447, in _pool_requests
    elems = r.get()
  File "C:\Python27\lib\multiprocessing\pool.py", line 567, in get
    raise self._value
ValueError

Line 281 is the second line from my code above with save(...).

It works fine if I comment out the attach line.

Any ideas?

Thanks,

rjwiii

**Comment by rjwiii** [comment]
The usual way:

vcn_appt = exchangelib.CalendarItem(folder=vcn_acct.calendar)

then add stuff with vcn.appt.subject, .start, .end, .required_attendees, .legacy_free_busy_status & .body ... then the two lines from above to attempt to attach the original e-mail and save/send ...

Sorry I can't post more code, but I'm in a corporate environment. If the above isn't clear enough, let me know and I will post as much code as I can without violating our security policy.

Thanks again for taking the time to help me.

rjwiii

**Comment by ecederstrand** [comment]
Thanks for the extra info. I'll have a look later today.

I found out the information about whether a response is required is probably carried in the `IsResponseRequested` field that we don't support yet.

**Comment by ecederstrand** [comment]
When you create a calendar item the 'start' and 'end'  fields are required. That's probably the source of your `ValueError` exception.

I sent myself a test email with the subject `Test for #239` and then had success with this test:

```python
from exchangelib import Account, Credentials, ItemAttachment, CalendarItem, UTC_NOW
from exchangelib.items import SEND_TO_ALL_AND_SAVE_COPY

a = Account(email, credentials=Credentials('email', 'password'), autodiscover=True)

vcn_email = a.inbox.get(subject='Test for #239')
vcn_appt = CalendarItem(
    folder=a.calendar, 
    subject=vcn_email.subject, 
    start=UTC_NOW(), 
    end=UTC_NOW() + timedelta(hours=2),
)
vcn_appt.attach(ItemAttachment(name=vcn_email.subject, item=vcn_email))
vcn_appt.save(send_meeting_invitations=SEND_TO_ALL_AND_SAVE_COPY)
```

If the `vcn_email` itself has a file attachment, you'll get a `ErrorRequiredPropertyMissing` unless you pull in [this commit](https://github.com/ecederstrand/exchangelib/commit/2804e013c639c7a13565e7b48dff528412c485d7).

**Comment by rjwiii** [comment]
I upgraded to 1.10.1 and tried it again ... It worked!

Now I just need to place it neatly in the appointment somewhere ...

Thanks.

Any idea when you will support the "IsResponseRequested" field?

rjwiii

**Comment by rjwiii** [comment]
Always seems to place the attachment between the last two lines of the meeting invite. Is there any control over the placement?

Thanks,

rjwiii

**Comment by ecederstrand** [comment]
I'm working through a patch to add support for most missing fields, among them `IsResponseRequested`. Should land within a week.

I have no idea about the placement, but you could try to add a link to the attachment in the HTML body, as described in https://github.com/ecederstrand/exchangelib#attachments

**Comment by rjwiii** [comment]
Is there a new version out? Can you provide an example in your docs?

Thanks,

rjwiii

**Comment by rjwiii** [comment]
is it in version 1.10.3?

**Comment by ecederstrand** [comment]
No, it's not released in a version yet. You'll have to install from GitHub for now.

**Comment by ecederstrand** [comment]
Patches available in v1.10.4. Closing

**Comment by rjwiii** [comment]
I tried it ... IT WORKED!

Thanks!
